File: src/agent/session.ts
import * as FileSystem from "@effect/platform/FileSystem";
import * as Path from "@effect/platform/Path";
import { Effect } from "effect";
import * as S from "effect/Schema";
import type { AgentConfig, AgentTurn } from "./loop.js";

const ToolCallSchema = S.Struct({
  id: S.String,
  name: S.String,
  arguments: S.String,
});

const ToolContentSchema = S.Union(
  S.Struct({ type: S.Literal("text"), text: S.String }),
  S.Struct({ type: S.Literal("image"), data: S.String, mimeType: S.String }),
);

const ToolResultSchema = S.Struct({
  toolCallId: S.String,
  name: S.String,
  result: S.Struct({ content: S.Array(ToolContentSchema) }),
  isError: S.Boolean,
});

const AgentTurnSchema = S.Struct({
  role: S.Union(S.Literal("assistant"), S.Literal("tool")),
  content: S.NullOr(S.String),
  toolCalls: S.optional(S.Array(ToolCallSchema)),
  toolResults: S.optional(S.Array(ToolResultSchema)),
});

const ChatMessageSchema = S.Struct({
  role: S.Union(S.Literal("system"), S.Literal("user"), S.Literal("assistant"), S.Literal("tool")),
  content: S.String,
  tool_call_id: S.optional(S.String),
  name: S.optional(S.String),
});

const SessionEventSchema = S.Union(
  S.Struct({
    type: S.Literal("session_start"),
    timestamp: S.String,
    sessionId: S.String,
    config: S.Struct({
      model: S.optional(S.String),
      systemPrompt: S.optional(S.String),
      maxTurns: S.optional(S.Number),
      temperature: S.optional(S.Number),
    }),
  }),
  S.Struct({
    type: S.Literal("user_message"),
    timestamp: S.String,
    content: S.String,
  }),
  S.Struct({
    type: S.Literal("turn"),
    timestamp: S.String,
    turn: AgentTurnSchema,
  }),
  S.Struct({
    type: S.Literal("message"),
    timestamp: S.String,
    message: ChatMessageSchema,
  }),
  S.Struct({
    type: S.Literal("session_end"),
    timestamp: S.String,
    totalTurns: S.Number,
    finalMessage: S.NullOr(S.String),
  }),
);

type SessionEvent = S.Schema.Type<typeof SessionEventSchema>;

export interface Session {
  id: string;
  config: AgentConfig;
  messages: Array<{ role: string; content: string; tool_call_id?: string; name?: string }>;
  turns: AgentTurn[];
  userMessage: string;
}

export class SessionError extends Error {
  readonly _tag = "SessionError";
  constructor(
    readonly reason: "not_found" | "parse_error" | "write_error",
    message: string,
  ) {
    super(message);
    this.name = "SessionError";
  }
}

const generateSessionId = () => {
  const now = new Date();
  const date = now.toISOString().slice(0, 10).replace(/-/g, "");
  const time = now.toISOString().slice(11, 19).replace(/:/g, "");
  const rand = Math.random().toString(36).slice(2, 6);
  return `session-${date}-${time}-${rand}`;
};

const timestamp = () => new Date().toISOString();

export const createSession = (
  config: AgentConfig,
  userMessage: string,
  sessionId?: string,
): Session => ({
  id: sessionId ?? generateSessionId(),
  config,
  messages: [],
  turns: [],
  userMessage,
});

export const appendEvent = (
  sessionPath: string,
  event: SessionEvent,
): Effect.Effect<void, SessionError, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const line = JSON.stringify(event) + "\n";
    yield* fs.writeFile(sessionPath, new TextEncoder().encode(line), { flag: "a" }).pipe(
      Effect.mapError((e) => new SessionError("write_error", `Failed to append event: ${e.message}`)),
    );
  });

export const writeSessionStart = (
  sessionPath: string,
  session: Session,
): Effect.Effect<void, SessionError, FileSystem.FileSystem> =>
  appendEvent(sessionPath, {
    type: "session_start",
    timestamp: timestamp(),
    sessionId: session.id,
    config: session.config,
  });

export const writeUserMessage = (
  sessionPath: string,
  content: string,
): Effect.Effect<void, SessionError, FileSystem.FileSystem> =>
  appendEvent(sessionPath, {
    type: "user_message",
    timestamp: timestamp(),
    content,
  });

export const writeTurn = (
  sessionPath: string,
  turn: AgentTurn,
): Effect.Effect<void, SessionError, FileSystem.FileSystem> =>
  appendEvent(sessionPath, {
    type: "turn",
    timestamp: timestamp(),
    turn: turn as S.Schema.Type<typeof AgentTurnSchema>,
  });

export const writeMessage = (
  sessionPath: string,
  message: { role: string; content: string; tool_call_id?: string; name?: string },
): Effect.Effect<void, SessionError, FileSystem.FileSystem> =>
  appendEvent(sessionPath, {
    type: "message",
    timestamp: timestamp(),
    message: message as S.Schema.Type<typeof ChatMessageSchema>,
  });

export const writeSessionEnd = (
  sessionPath: string,
  totalTurns: number,
  finalMessage: string | null,
): Effect.Effect<void, SessionError, FileSystem.FileSystem> =>
  appendEvent(sessionPath, {
    type: "session_end",
    timestamp: timestamp(),
    totalTurns,
    finalMessage,
  });

export const loadSession = (
  sessionPath: string,
): Effect.Effect<Session, SessionError, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;

    const exists = yield* fs.exists(sessionPath).pipe(
      Effect.mapError((e) => new SessionError("not_found", `Cannot check file: ${e.message}`)),
    );
    if (!exists) {
      return yield* Effect.fail(new SessionError("not_found", `Session file not found: ${sessionPath}`));
    }

    const content = yield* fs.readFileString(sessionPath).pipe(
      Effect.mapError((e) => new SessionError("not_found", `Cannot read file: ${e.message}`)),
    );

    const lines = content.trim().split("\n").filter((l) => l.length > 0);
    if (lines.length === 0) {
      return yield* Effect.fail(new SessionError("parse_error", "Empty session file"));
    }

    let session: Session | null = null;

    for (const line of lines) {
      const parsed = yield* Effect.try({
        try: () => JSON.parse(line) as SessionEvent,
        catch: () => new SessionError("parse_error", `Invalid JSON: ${line.slice(0, 50)}...`),
      });

      const event = yield* S.decodeUnknown(SessionEventSchema)(parsed).pipe(
        Effect.mapError(() => new SessionError("parse_error", `Invalid event format: ${line.slice(0, 50)}...`)),
      );

      switch (event.type) {
        case "session_start":
          session = {
            id: event.sessionId,
            config: {
              ...(event.config.model ? { model: event.config.model } : {}),
              ...(event.config.systemPrompt ? { systemPrompt: event.config.systemPrompt } : {}),
              ...(event.config.maxTurns !== undefined ? { maxTurns: event.config.maxTurns } : {}),
              ...(event.config.temperature !== undefined ? { temperature: event.config.temperature } : {}),
            },
            messages: [],
            turns: [],
            userMessage: "",
          };
          break;

        case "user_message":
          if (session) {
            session.userMessage = event.content;
          }
          break;

        case "message":
          if (session) {
            const msg: { role: string; content: string; tool_call_id?: string; name?: string } = {
              role: event.message.role,
              content: event.message.content,
            };
            if (event.message.tool_call_id) msg.tool_call_id = event.message.tool_call_id;
            if (event.message.name) msg.name = event.message.name;
            session.messages.push(msg);
          }
          break;

        case "turn":
          if (session) {
            session.turns.push(event.turn as AgentTurn);
          }
          break;

        case "session_end":
          break;
      }
    }

    if (!session) {
      return yield* Effect.fail(new SessionError("parse_error", "No session_start event found"));
    }

    return session;
  });

export const getSessionPath = (
  sessionsDir: string,
  sessionId: string,
): Effect.Effect<string, never, Path.Path> =>
  Effect.gen(function* () {
    const path = yield* Path.Path;
    return path.join(sessionsDir, `${sessionId}.jsonl`);
  });

export const listSessions = (
  sessionsDir: string,
): Effect.Effect<string[], SessionError, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;

    const exists = yield* fs.exists(sessionsDir).pipe(
      Effect.mapError((e) => new SessionError("not_found", e.message)),
    );
    if (!exists) {
      return [];
    }

    const entries = yield* fs.readDirectory(sessionsDir).pipe(
      Effect.mapError((e) => new SessionError("not_found", e.message)),
    );

    return entries
      .filter((e) => e.endsWith(".jsonl"))
      .map((e) => e.replace(".jsonl", ""))
      .sort()
      .reverse();
  });

File: crates/agent/src/session.rs
//! Session management for agent runs
//!
//! Sessions are stored as JSONL files with events for each action.
//! This allows for crash recovery and replay.

use crate::error::{AgentError, AgentResult};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use std::fs::{self, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

// ============================================================================
// Session Event Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub id: String,
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ToolContent {
    Text { text: String },
    Image { data: String, mime_type: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolResult {
    pub tool_call_id: String,
    pub name: String,
    pub result: ToolResultContent,
    pub is_error: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResultContent {
    pub content: Vec<ToolContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentTurn {
    pub role: TurnRole,
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<ToolCall>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_results: Option<Vec<ToolResult>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TurnRole {
    Assistant,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageRole {
    System,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatMessage {
    pub role: MessageRole,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_prompt: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SessionEvent {
    SessionStart {
        timestamp: String,
        #[serde(rename = "sessionId")]
        session_id: String,
        config: SessionConfig,
    },
    UserMessage {
        timestamp: String,
        content: String,
    },
    Turn {
        timestamp: String,
        turn: AgentTurn,
    },
    Message {
        timestamp: String,
        message: ChatMessage,
    },
    SessionEnd {
        timestamp: String,
        #[serde(rename = "totalTurns")]
        total_turns: u32,
        #[serde(rename = "finalMessage")]
        final_message: Option<String>,
    },
    LogTrimmed {
        timestamp: String,
        dropped: u32,
        kept: u32,
        reason: String,
    },
}

// ============================================================================
// Session
// ============================================================================

#[derive(Debug, Clone)]
pub struct Session {
    pub id: String,
    pub config: SessionConfig,
    pub messages: Vec<ChatMessage>,
    pub turns: Vec<AgentTurn>,
    pub user_message: String,
}

impl Session {
    /// Create a new session
    pub fn new(config: SessionConfig, user_message: impl Into<String>) -> Self {
        Self {
            id: generate_session_id(),
            config,
            messages: Vec::new(),
            turns: Vec::new(),
            user_message: user_message.into(),
        }
    }

    /// Create a session with a specific ID
    pub fn with_id(
        id: impl Into<String>,
        config: SessionConfig,
        user_message: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            config,
            messages: Vec::new(),
            turns: Vec::new(),
            user_message: user_message.into(),
        }
    }
}

// ============================================================================
// Session I/O
// ============================================================================

/// Configuration for session log trimming
#[derive(Debug, Clone)]
pub struct TrimConfig {
    /// Maximum size in bytes before trimming
    pub max_bytes: u64,
    /// Maximum number of lines before trimming
    pub max_lines: usize,
    /// Number of lines to keep at the start (preserve session_start)
    pub keep_head: usize,
    /// Number of lines to keep at the end (preserve recent events)
    pub keep_tail: usize,
}

impl Default for TrimConfig {
    fn default() -> Self {
        Self {
            max_bytes: 10 * 1024 * 1024, // 10 MB
            max_lines: 10_000,
            keep_head: 10,
            keep_tail: 500,
        }
    }
}

/// Append an event to the session log
pub fn append_event(session_path: &Path, event: &SessionEvent) -> AgentResult<()> {
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(session_path)?;

    let line = serde_json::to_string(event)? + "\n";
    file.write_all(line.as_bytes())?;

    // Check if trimming is needed
    maybe_trim_session_log(session_path, &TrimConfig::default())?;

    Ok(())
}

/// Write session start event
pub fn write_session_start(session_path: &Path, session: &Session) -> AgentResult<()> {
    append_event(
        session_path,
        &SessionEvent::SessionStart {
            timestamp: timestamp(),
            session_id: session.id.clone(),
            config: session.config.clone(),
        },
    )
}

/// Write user message event
pub fn write_user_message(session_path: &Path, content: impl Into<String>) -> AgentResult<()> {
    append_event(
        session_path,
        &SessionEvent::UserMessage {
            timestamp: timestamp(),
            content: content.into(),
        },
    )
}

/// Write turn event
pub fn write_turn(session_path: &Path, turn: &AgentTurn) -> AgentResult<()> {
    append_event(
        session_path,
        &SessionEvent::Turn {
            timestamp: timestamp(),
            turn: turn.clone(),
        },
    )
}

/// Write message event
pub fn write_message(session_path: &Path, message: &ChatMessage) -> AgentResult<()> {
    append_event(
        session_path,
        &SessionEvent::Message {
            timestamp: timestamp(),
            message: message.clone(),
        },
    )
}

/// Write session end event
pub fn write_session_end(
    session_path: &Path,
    total_turns: u32,
    final_message: Option<String>,
) -> AgentResult<()> {
    append_event(
        session_path,
        &SessionEvent::SessionEnd {
            timestamp: timestamp(),
            total_turns,
            final_message,
        },
    )
}

/// Load a session from a JSONL file
pub fn load_session(session_path: &Path) -> AgentResult<Session> {
    if !session_path.exists() {
        return Err(AgentError::Session(format!(
            "Session file not found: {}",
            session_path.display()
        )));
    }

    let file = fs::File::open(session_path)?;
    let reader = BufReader::new(file);

    let mut session: Option<Session> = None;

    for line in reader.lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }

        let event: SessionEvent = serde_json::from_str(&line).map_err(|e| {
            AgentError::Session(format!("Invalid JSON in session file: {}", e))
        })?;

        match event {
            SessionEvent::SessionStart {
                session_id,
                config,
                ..
            } => {
                session = Some(Session {
                    id: session_id,
                    config,
                    messages: Vec::new(),
                    turns: Vec::new(),
                    user_message: String::new(),
                });
            }
            SessionEvent::UserMessage { content, .. } => {
                if let Some(ref mut s) = session {
                    s.user_message = content;
                }
            }
            SessionEvent::Message { message, .. } => {
                if let Some(ref mut s) = session {
                    s.messages.push(message);
                }
            }
            SessionEvent::Turn { turn, .. } => {
                if let Some(ref mut s) = session {
                    s.turns.push(turn);
                }
            }
            SessionEvent::LogTrimmed { .. } | SessionEvent::SessionEnd { .. } => {
                // No action needed
            }
        }
    }

    session.ok_or_else(|| AgentError::Session("No session_start event found".to_string()))
}

/// Get the path for a session file
pub fn get_session_path(sessions_dir: &Path, session_id: &str) -> PathBuf {
    sessions_dir.join(format!("{}.jsonl", session_id))
}

/// List all session IDs in a directory
pub fn list_sessions(sessions_dir: &Path) -> AgentResult<Vec<String>> {
    if !sessions_dir.exists() {
        return Ok(Vec::new());
    }

    let mut sessions = Vec::new();
    for entry in fs::read_dir(sessions_dir)? {
        let entry = entry?;
        let path = entry.path();
        if let Some(ext) = path.extension() {
            if ext == "jsonl" {
                if let Some(stem) = path.file_stem() {
                    if let Some(name) = stem.to_str() {
                        sessions.push(name.to_string());
                    }
                }
            }
        }
    }

    sessions.sort();
    sessions.reverse();
    Ok(sessions)
}

// ============================================================================
// Helpers
// ============================================================================

fn generate_session_id() -> String {
    let now = Utc::now();
    let date = now.format("%Y%m%d").to_string();
    let time = now.format("%H%M%S").to_string();
    let rand: String = (0..4)
        .map(|_| {
            let idx = rand::random::<usize>() % 36;
            if idx < 10 {
                (b'0' + idx as u8) as char
            } else {
                (b'a' + (idx - 10) as u8) as char
            }
        })
        .collect();
    format!("session-{}-{}-{}", date, time, rand)
}

fn timestamp() -> String {
    Utc::now().to_rfc3339()
}

/// Trim a session log if it exceeds size limits
fn maybe_trim_session_log(session_path: &Path, config: &TrimConfig) -> AgentResult<()> {
    let metadata = fs::metadata(session_path)?;

    // Check if trimming is needed based on file size
    if metadata.len() < config.max_bytes {
        return Ok(());
    }

    let content = fs::read_to_string(session_path)?;
    let lines: Vec<&str> = content.lines().filter(|l| !l.trim().is_empty()).collect();

    // Check if trimming is needed based on line count
    if lines.len() <= config.max_lines {
        return Ok(());
    }

    // Trim the log, keeping head and tail
    let total_keep = config.keep_head + config.keep_tail;
    if lines.len() <= total_keep {
        return Ok(());
    }

    let dropped = lines.len() - total_keep;
    let mut new_lines: Vec<String> = Vec::with_capacity(total_keep + 1);

    // Keep head
    new_lines.extend(lines.iter().take(config.keep_head).map(|s| s.to_string()));

    // Add trim marker
    let trim_event = SessionEvent::LogTrimmed {
        timestamp: timestamp(),
        dropped: dropped as u32,
        kept: total_keep as u32,
        reason: "exceeded size limit".to_string(),
    };
    new_lines.push(serde_json::to_string(&trim_event).unwrap_or_default());

    // Keep tail
    new_lines.extend(lines.iter().skip(lines.len() - config.keep_tail).map(|s| s.to_string()));

    let new_content = new_lines.join("\n") + "\n";
    fs::write(session_path, new_content)?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_session_creation() {
        let config = SessionConfig {
            model: Some("claude-3".to_string()),
            ..Default::default()
        };
        let session = Session::new(config, "Hello, world!");

        assert!(session.id.starts_with("session-"));
        assert_eq!(session.user_message, "Hello, world!");
        assert!(session.turns.is_empty());
        assert!(session.messages.is_empty());
    }

    #[test]
    fn test_session_roundtrip() -> AgentResult<()> {
        let dir = tempdir()?;
        let session_path = dir.path().join("test-session.jsonl");

        // Create and write session
        let config = SessionConfig {
            model: Some("claude-3".to_string()),
            max_turns: Some(10),
            ..Default::default()
        };
        let session = Session::with_id("test-123", config, "Test message");

        write_session_start(&session_path, &session)?;
        write_user_message(&session_path, "Hello!")?;
        write_message(
            &session_path,
            &ChatMessage {
                role: MessageRole::Assistant,
                content: "Hi there!".to_string(),
                tool_call_id: None,
                name: None,
            },
        )?;
        write_session_end(&session_path, 1, Some("Goodbye".to_string()))?;

        // Load and verify
        let loaded = load_session(&session_path)?;

        assert_eq!(loaded.id, "test-123");
        assert_eq!(loaded.config.model, Some("claude-3".to_string()));
        assert_eq!(loaded.user_message, "Hello!");
        assert_eq!(loaded.messages.len(), 1);
        assert_eq!(loaded.messages[0].content, "Hi there!");

        Ok(())
    }

    #[test]
    fn test_list_sessions() -> AgentResult<()> {
        let dir = tempdir()?;

        // Create some session files
        fs::write(dir.path().join("session-a.jsonl"), "")?;
        fs::write(dir.path().join("session-b.jsonl"), "")?;
        fs::write(dir.path().join("session-c.jsonl"), "")?;
        fs::write(dir.path().join("not-a-session.txt"), "")?;

        let sessions = list_sessions(dir.path())?;

        assert_eq!(sessions.len(), 3);
        // Sorted in reverse order
        assert_eq!(sessions[0], "session-c");
        assert_eq!(sessions[1], "session-b");
        assert_eq!(sessions[2], "session-a");

        Ok(())
    }

    #[test]
    fn test_session_event_serialization() {
        let event = SessionEvent::SessionStart {
            timestamp: "2025-01-01T00:00:00Z".to_string(),
            session_id: "sess-123".to_string(),
            config: SessionConfig::default(),
        };

        let json = serde_json::to_string(&event).unwrap();
        assert!(json.contains("\"type\":\"session_start\""));
        assert!(json.contains("\"sessionId\":\"sess-123\""));

        // Roundtrip
        let parsed: SessionEvent = serde_json::from_str(&json).unwrap();
        match parsed {
            SessionEvent::SessionStart { session_id, .. } => {
                assert_eq!(session_id, "sess-123");
            }
            _ => panic!("Wrong event type"),
        }
    }
}
